server/src/main/java/com/thoughtworks/go/server/datamigration/M001.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2018 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.datamigration;
+
+import com.google.gson.Gson;
+import org.apache.commons.lang3.StringUtils;
+
+import java.sql.*;
+
+class M001 {
+    private static final String ID = "id";
+    private static final String SELECTIONS = "selections";
+    private static final String BLACKLIST = "isblacklist";
+    private static final int SCHEMA = 1;
+    private static final Gson GSON = new Gson();
+
+    static Migration convertPipelineSelectionsToFilters() {
+        return (cxn) -> {
+            if (!required(cxn)) return;
+
+            try (Statement s = cxn.createStatement()) {
+                final ResultSet rs = s.executeQuery("SELECT id, selections, isblacklist FROM pipelineselections WHERE version = 0");
+                while (rs.next()) {
+                    perform(cxn, rs.getLong(ID), rs.getString(SELECTIONS), rs.getBoolean(BLACKLIST));
+                }
+            }
+        };
+    }
+
+    static void perform(Connection cxn, long id, String selections, boolean isBlacklist) throws SQLException {
+        try (PreparedStatement ps = cxn.prepareStatement("UPDATE pipelineselections SET selections = NULL, version = ?, filters = ? WHERE id = ?")) {
+            ps.setInt(1, SCHEMA);
+            ps.setString(2, asJson(selections, isBlacklist));
+            ps.setLong(3, id);
+            ps.executeUpdate();
```

Comment:
The data migration will essentially convert the pipeline selections representation into a JSON document
with this structure:

```
{
  filters: [
    { name: "first filter", type: "whitelist", pipelines: [ "p1", "p2", ..., "pN" ] },
    { name: "second filter", type: "blacklist", pipelines: [ "p4", "p5" ] },
    ...
  ]
}
```

The implications here are that we can support multiple pipeline filters per user as a single row in the
table and that this will allow us to consolidate and shift more logic to the frontend. The backend can
be written in a way such that its only concerns are serialization, deserialization, and persistence.